gtsfm/scene_optimizer.py

```diff
@@ -119,14 +126,27 @@ def create_computation_graph(
             keypoints_graph_list += [delayed_dets]
             descriptors_graph_list += [delayed_descs]
 
-        # estimate two-view geometry and get indices of verified correspondences.
+        # Add detections and descriptors to the Retriever and create local matching graph.
+        local_matching_graph_dict = self.image_retriever.create_computation_graph(
```

Comment:
It might make more sense to have the Retriever run after the feature extractor, and then have it spit out potential pairs, and then just keep the existing for loop over pairs (each one goes into TwoViewEstimator).

So I was thinking TwoViewEstimator would not be an attribute of the retriever (I think retriever is more minor and TwoViewEstimator is more major in class prominence)

Reply:
Is the method that you described parallelizable? Before we call the TwoViewEstimator, we have to know the image pair indices `i1` and `i2`. 

I believe the way I've coded it here allows for the pair retrieval to be delayed and parallelized. Although, I may be wrong.

Reply:
Can retrieval be parallelized in the way that you describe? Before we call the TwoViewEstimator, the image indices `i1` and `i2` have to already have been computed.

I believe making delayed calls to the TwoViewEstimator from the Retriever allows for the pair proposal process to be done in parallel. Otherwise, we have to evaluate all of the features first and pass into the Retriever, creating a bottleneck. I'm not 100% sure on this though.

However, for the case when we're considering global descriptors, I'm not sure this bottleneck can be avoided (as we need to compute the global descriptors for all the images first before proposing image pairs for the TwoViewEstimator).

Reply:
@johnwlambert Still wanted to get your opinion on this. The way I see it there are three options since the TwoViewEstimator will be called based on the output of the Retriever:

1. Make the TwoViewEstimator an attribute of the Retriever and only call it if the Retriever decides it's appropriate (this is what's implemented in this branch currently)
2. Pass the (Delayed) output of the Retriever to the TwoViewEstimator and let the estimator evaluate the output
3. Don't use any delayed operations in the Retriever and evaluate its output in the main loop of the the SceneOptimizer, which would explicitly create a bottleneck between feature extraction and the TwoViewEstimator

Reply:
Hi Travis, thanks for following up on this. I like option #2 most.